internal/staging_lockfile/staging_lockfile.go

```diff
@@ -0,0 +1,145 @@
+package staging_lockfile
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/containers/storage/internal/rawfilelock"
+)
+
+// StagingLockFile represents a file lock used to coordinate access to staging areas.
+// Typical usage is via CreateAndLock or TryLockPath, both of which return a StagingLockFile
+// that must eventually be released with UnlockAndDelete. This ensures that access
+// to the staging file is properly synchronized both within and across processes.
+//
+// WARNING: This struct MUST NOT be created manually. Use the provided helper functions instead.
+type StagingLockFile struct {
+	// Locking invariant: If stagingLockFileLock is not locked, a StagingLockFile for a particular
+	// path exists if the current process currently owns the lock for that file, and it is recorded in stagingLockFiles.
+	//
+	// The following fields are only set when constructing *StagingLockFile, and must never be modified afterwards.
+	// They are safe to access without any other locking.
```

Comment:
This is incorrect:
- `file` is modified by `UnlockAndDelete`
- (Pedantic `fd` is not modified, but we close the handle, so it eventually becomes invalid.)

Actually nowadays we never access these fields from a goroutine that doesn’t own the lock (if we find an existing entry in `tryAcquireLockForFile`, we don’t even read it); so “these fields can only be accessed by the goroutine owning the lock” seems fine.